lib/keystore.cc

```diff
@@ -82,55 +131,112 @@ rpmRC keystore_fs::delete_key(rpmtxn txn, rpmPubkey key)
 rpmRC keystore_fs::import_key(rpmtxn txn, rpmPubkey key, int replace, rpmFlags flags)
 {
     rpmRC rc = RPMRC_FAIL;
-    const char *fp = rpmPubkeyFingerprintAsHex(key);
-    char *keyfmt = rstrscat(NULL, "gpg-pubkey-", fp, ".key", NULL);
-    char *keyval = rpmPubkeyArmorWrap(key);
-    char *path = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", keyfmt);
-    char *tmppath = NULL;
+    string fp = rpmPubkeyFingerprintAsHex(key);
+    string keyfmt = "gpg-pubkey-" + fp + ".key";
+    char *dir = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", NULL);
+    string dirstr = dir;
 
-    if (replace) {
-	rasprintf(&tmppath, "%s.new", path);
-	unlink(tmppath);
+    rc = write_key_to_disk(key, dirstr, keyfmt, replace, flags);
+
+    if (!rc && replace) {
+	/* find and delete the old pubkey entry */
+	if (delete_key(txn, fp, keyfmt) == RPMRC_NOTFOUND) {
+	    /* make sure an old, short keyid version gets removed */
+	    delete_key(txn, fp.substr(32), keyfmt);
+	}
     }
 
-    if (rpmMkdirs(rpmtxnRootDir(txn), "%{_keyringpath}")) {
+    free(dir);
+    return rc;
+}
+
+/*****************************************************************************/
+
+static int acquire_write_lock(rpmtxn txn)
+{
+    char * keyringpath = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", NULL);
+    char * lockpath = rpmGenPath(rpmtxnRootDir(txn), "%{_keyringpath}/", "writelock");
+    int fd = -1;
+
+    if (rpmMkdirs(NULL, keyringpath)) {
 	rpmlog(RPMLOG_ERR, _("failed to create keyring directory %s: %s\n"),
-		path, strerror(errno));
+	       keyringpath, strerror(errno));
 	goto exit;
     }
 
-    if (FD_t fd = Fopen(tmppath ? tmppath : path, "wx")) {
-	size_t keylen = strlen(keyval);
-	if (Fwrite(keyval, 1, keylen, fd) == keylen)
-	    rc = RPMRC_OK;
-	Fclose(fd);
+    if ((fd = open(lockpath, O_WRONLY|O_CREAT)) == -1) {
+	rpmlog(RPMLOG_ERR, _("Can't create writelock for keyring at %s: %s\n"), keyringpath, strerror(errno));
+    } else if (flock(fd, LOCK_EX|LOCK_NB)) {
+	rpmlog(RPMLOG_ERR, _("Can't acquire writelock for keyring at %s\n"), keyringpath);
+	close(fd);
+	fd = -1;
     }
-    if (!rc && tmppath && rename(tmppath, path) != 0)
-	rc = RPMRC_FAIL;
 
-    if (rc) {
-	rpmlog(RPMLOG_ERR, _("failed to import key: %s: %s\n"),
-		tmppath ? tmppath : path, strerror(errno));
-	if (tmppath)
-	    unlink(tmppath);
-    }
+ exit:
+    free(keyringpath);
+    free(lockpath);
+    return fd;
+}
 
-    if (!rc && replace) {
-	/* find and delete the old pubkey entry */
-	if (delete_key(txn, fp, keyfmt) == RPMRC_NOTFOUND) {
-	    /* make sure an old, short keyid version gets removed */
-	    delete_key(txn, fp+32, keyfmt);
-	}
-    }
+static void free_write_lock(int fd)
+{
+    flock(fd, LOCK_UN);
```

Comment:
I wonder if we shouldn't also remove the `writelock` file here.

Reply:
Hmm, perhaps not really. The spec only says the store should be assumed unlocked if the file doesn't exist, but its presence doesn't really matter (as long as we use `flock()` of course).

Reply:
OK, you even asked the same question above (just noticed)...

Reply:
I'd say it's fine to keep the file around, whoever wants to access the keystore is expected to use the file locking syscall anyway.